openapi3/schema.go

```diff
@@ -1231,3 +1273,113 @@ func isSliceOfUniqueItems(xs []interface{}) bool {
 func unsupportedFormat(format string) error {
 	return fmt.Errorf("Unsupported 'format' value '%s'", format)
 }
+
+func invalidDefaultValue(s *Schema, err error) error {
+	return fmt.Errorf("default value error when validating schema %s (default: %v, error: %s)", "*name*", s.Default, err.Error())
+}
+
+func validateDefaultProperty(s *Schema, f func(s *Schema) error) error {
+	if s.Default == nil {
+		return nil
+	}
+	if err := f(s); err != nil {
+		return invalidDefaultValue(s, err)
+	}
+	return nil
+}
+
+func validateStringDefaultProperty(s *Schema, f func(s *Schema) error) error {
+	if s.Default == nil {
+		return nil
+	}
+	_, ok := s.Default.(string)
+	if !ok {
+		return invalidDefaultValue(s, errors.New("not a string"))
+	}
+	if err := f(s); err != nil {
+		return invalidDefaultValue(s, err)
+	}
+	return nil
+}
+
+func validateNumberDefaultProperty(s *Schema, f func(s *Schema) error) error {
+	if s.Default == nil {
+		return nil
+	}
+	_, ok := s.Default.(float64)
+	if !ok {
+		return invalidDefaultValue(s, errors.New("not a number"))
+	}
+	if err := f(s); err != nil {
+		return invalidDefaultValue(s, err)
+	}
+	return nil
+}
+
+func dfltStr(s *Schema) string { return fmt.Sprintf("%v", s.Default) }
+
+func validateBooleanDefault(s *Schema) (err error) {
+	return validateDefaultProperty(s, func(s *Schema) (err error) { _, err = strconv.ParseBool(dfltStr(s)); return err })
+}
+
+func validateNumberDefault(s *Schema) (err error) {
+	return validateNumberDefaultProperty(s, func(s *Schema) (err error) { _, err = strconv.ParseFloat(dfltStr(s), 64); return err })
+}
+
+func validateFloatDefault(s *Schema) error {
+	return validateNumberDefaultProperty(s, func(s *Schema) (err error) { _, err = strconv.ParseFloat(dfltStr(s), 32); return err })
+}
+
+func validateDoubleDefault(s *Schema) error {
+	return validateNumberDefaultProperty(s, func(s *Schema) (err error) { return nil })
+}
+
+func validateIntegerDefault(s *Schema) error {
+	return validateInt64Default(s)
+}
+
+func validateInt32Default(s *Schema) error {
+	return validateNumberDefaultProperty(s, func(s *Schema) (err error) {
+		f := s.Default.(float64)
+		if f != float64(int32(f)) || math.MinInt32 > f || math.MaxInt32 < f {
+			return errors.New("number not an int or out of range (int32)")
```

Comment:
Or is the enemy of debugging IMHO.  The cast check and the boundary checks should probably be separate.

Reply:
addressed